// scraper-js/packages/core/dist/scrape.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.scrape = void 0;
const axios_1 = __importDefault(require("axios"));
const image_1 = require("./image");
const parse_1 = require("./parse");
function scrape(url, options = {
    client: axios_1.default.create({
        timeout: 1000,
    }),
}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { client } = options;
        const response = yield client.get(url);
        const result = yield (0, parse_1.parse)(response.data, {
            parser: options.parser,
        });
        return {
            featureImage: (0, image_1.featureImageFromParseResult)(result, {
                rankImage: options.rankImage,
            }),
            images: (0, image_1.imagesFromParseResult)(result, {
                rankImage: options.rankImage,
            }),
            html: result.document.html(),
            linkedData: result.linkedData,
            openGraph: result.openGraph,
        };
    });
}
exports.scrape = scrape;
